src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeReader.java

```diff
@@ -0,0 +1,455 @@
+            this.visitor = visitor;
+            this.packedValue = new byte[packedBytesLength];
+            this.origIndex = new short[maxPointsInLeafNode];
+        }
+
+        @Override
+        public void executeInternal(final PriorityQueue<PeekablePostingList> postingLists) throws IOException
+        {
+            collectPostingLists(postingLists, minPackedValue, maxPackedValue);
+        }
+
+        private void collectPostingLists(PriorityQueue<PeekablePostingList> postingLists, byte[] cellMinPacked, byte[] cellMaxPacked) throws IOException
+        {
+            context.checkpoint();
+
+            final Relation r = visitor.compare(cellMinPacked, cellMaxPacked);
+
+            if (r == Relation.CELL_OUTSIDE_QUERY)
+            {
+                // This cell is fully outside the query shape: stop recursing
+                return;
+            }
+
+            if (r == Relation.CELL_INSIDE_QUERY)
+            {
+                // This cell is fully inside the query shape: recursively add all points in this cell without filtering
+                super.collectPostingLists(postingLists);
+                return;
+            }
+
+            if (index.isLeafNode())
+            {
+                if (index.nodeExists())
+                    filterLeaf(postingLists);
+                return;
+            }
+
+            visitNode(postingLists, cellMinPacked, cellMaxPacked);
+        }
+
+        private void filterLeaf(PriorityQueue<PeekablePostingList> postingLists) throws IOException
+        {
+            treeInput.seek(index.getLeafBlockFP());
+
+            int count = treeInput.readVInt();
+
+            // loading doc ids occurred here prior
+
+            int orderMapLength = treeInput.readVInt();
+
+            long orderMapPointer = treeInput.getFilePointer();
+
+            SeekingRandomAccessInput randomAccessInput = new SeekingRandomAccessInput(treeInput);
+            for (int x = 0; x < count; x++)
+            {
+                origIndex[x] = (short) LeafOrderMap.getValue(randomAccessInput, orderMapPointer, x, leafOrderMapReader);
+            }
+
+            // seek beyond the ordermap
+            treeInput.seek(orderMapPointer + orderMapLength);
+
+            FixedBitSet fixedBitSet = visitDocValues(treeInput, count, visitor, origIndex);
+
+            int nodeID = index.getNodeID();
+
+            if (postingsIndex.exists(nodeID) && fixedBitSet.cardinality() > 0)
+            {
+                long pointer = postingsIndex.getPostingsFilePointer(nodeID);
+                postingLists.add(initFilteringPostingReader(pointer, fixedBitSet));
+            }
+        }
+
+        void visitNode(PriorityQueue<PeekablePostingList> postingLists, byte[] cellMinPacked, byte[] cellMaxPacked) throws IOException
+        {
+            byte[] splitPackedValue = index.getSplitPackedValue();
+            BytesRef splitDimValue = index.getSplitDimValue();
```

Comment:
nit: I'm puzzling about the `splitDimValue` naming here. We've only got one dimension, so does `splitValuesStack` need to be a two-dimensional array? (i.e. Is there only one split value per level?)

Reply:
Hmm, so this opened a bit of a can of worms. I need to look at this in much more detail. I removed both of these from the method and from the `TraversalState` (was `PackedIndexTree`) and all the tests still pass. 

As I look at it, it seems that most of this was in place to handle the splitting around multiple dimensions and because we don't need them. We can just pass the min/max packed values through the recursion because the split dimension never changes. 

If this is the case we don't need to store some of this information in the in-memory state which could significantly reduce the in-memory requirement for the indexes.

I need to do a bit more investigation of this to make sure that I'm not missing anything.

Reply:
I have got to the bottom of this part of the code and have established that we don't need the `splitPackedValuesStack` and I have removed this part of the code. I have also removed all the copying forward and backward between different values as we recursed. AFAICT this was done because in the KDTree we were moving over an n-dimensional space rather than a single dimension. Overall, this has made the traversing code a lot simpler.

I now have an understanding of the balancing operation and of the overall structure of the resulting tree. I am just working through a few final edge conditions and I will get some documentation in place for these methods.